### PR TITLE
CRNS-265: Bidirectional scroll fixes for android

### PR DESCRIPTION
**Pull request by vishalnarkhede**
# Submit a pull request

## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] The code changes follow best practices
- [ ] Code changes are tested (add some information if not applicable)

### CODE REVIEW

File: src/components/MessageList/MessageList.tsx [hunk 1]
```diff
@@ -718,22 +820,29 @@ const MessageListWithContext = <
             )}
           </View>
         }
-        ListFooterComponent={FooterComponent}
+        ListFooterComponent={() =>
```
Comment (nhannah) [inline src/components/MessageList/MessageList.tsx:823]: onEndReachedInProgress I think it makes more sense to have a default FooterComponent then only hsouw it when onEndReachedInProgress? Or some logic that let's a user basically replace ours but still have access to the logic, e.g. `() => <FotterComponent onEndReachedInprogress />`

File: src/components/MessageList/MessageList.tsx [hunk 2]
```diff
@@ -718,22 +820,29 @@ const MessageListWithContext = <
...
+            FooterComponent
+          ) : onEndReachedInProgress ? (
+            <View style={[styles.activityIndicatorContainer]}>
+              <ActivityIndicator color={accent_blue} size='small' />
+            </View>
+          ) : null
+        }
         // TODO: Scrolling doesn't work perfectly with this loading indicator. Investigate and fix.
         ListHeaderComponent={() =>
           HeaderComponent ? (
```
Comment (nhannah) [inline src/components/MessageList/MessageList.tsx:834]: Same for Header Component

File: src/components/MessageList/MessageList.tsx [hunk 3]
```diff
@@ -563,6 +588,75 @@ const MessageListWithContext = <
     return null;
   };
 
+  const maybeCallOnStartReached = () => {
```
Comment (nhannah) [inline src/components/MessageList/MessageList.tsx:591]: I think you need to explain what the goal is here, I get what you are doing, and pretty sure we can avoid casting things to promises, but at first look this seems hacky so I am trying to understand why it's needed?

**Review by nhannah — CHANGES_REQUESTED** [review]

File: src/components/Channel/Channel.tsx [hunk 4]
```diff
@@ -708,7 +708,10 @@ const ChannelWithContext = <
         ) {
           const mostRecentMessage =
             channel.state.messages[channel.state.messages.length - 1];
-          await queryAfterMessage(mostRecentMessage.id, 5);
+          await queryAfterMessage(
+            mostRecentMessage.id,
+            10 - channel.state.messages.length,
```
Comment (virdesai) [inline src/components/Channel/Channel.tsx:713]: this could become negative?

File: src/components/MessageList/MessageList.tsx [hunk 5]
```diff
@@ -718,22 +820,29 @@ const MessageListWithContext = <
             )}
           </View>
         }
-        ListFooterComponent={FooterComponent}
+        ListFooterComponent={() =>
+          FooterComponent ? (
+            FooterComponent
+          ) : onEndReachedInProgress ? (
+            <View style={[styles.activityIndicatorContainer]}>
```
Comment (virdesai) [inline src/components/MessageList/MessageList.tsx:827]:
```suggestion
            <View style={styles.activityIndicatorContainer}>
```

File: src/components/MessageList/MessageList.tsx (hunk 3 above)
Comment (virdesai) [reply src/components/MessageList/MessageList.tsx:591]: ya looks like a very specific implementation that could be fragile

Comment (vishalnarkhede) [reply src/components/MessageList/MessageList.tsx:591]: > so I am trying to understand why it's needed

You mean why we need this `maybe` part?

Comment (vishalnarkhede) [reply src/components/MessageList/MessageList.tsx:591]: > I think you need to explain what the goal is here

Can do!!

File: src/components/Channel/Channel.tsx (hunk 4 above)
Comment (vishalnarkhede) [reply src/components/Channel/Channel.tsx:713]: Nope, it's wrapped under `if` condition!! All of this complex logic is mainly for android-the-wierdo

File: src/components/MessageList/MessageList.tsx (hunk 1 above)
Comment (vishalnarkhede) [reply src/components/MessageList/MessageList.tsx:823]: Hmm good point indeed. I will go with the first approach  (`to have a default FooterComponent`).

File: src/components/MessageList/InlineLoadingIndicator.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+  if (loadingMore) {
+    return (
+      <View style={styles.activityIndicatorContainer}>
+        <ActivityIndicator color={accent_blue} size='small' />
+      </View>
+    );
+  }
+
+  return null;
```
Comment (nhannah) [inline src/components/MessageList/InlineLoadingIndicator.tsx:33]: Could be like swift and do the early return, i.e. `if (!loadingMore) { return null }

File: src/components/MessageList/MessageList.tsx [hunk 7]
```diff
@@ -64,12 +63,12 @@ import type {
   DefaultUserType,
   UnknownType,
 } from '../../types/types';
+import {
```
Comment (nhannah) [inline src/components/MessageList/MessageList.tsx:66]: Import order should move this up

**Review by nhannah — APPROVED** [review]

**Pull request merged**